### DISCUSSION TITLE
Overwhelming

### DISCUSSION THREAD

**Discussion by Rand-Random** [dotnet/maui#122] [General]
Maybe I am not alone with the feeling, but the current situation is a bit overwhelming and hard to pick the right choice.

Everytime a new idea/product pop ups, the feeling increases.

There also seems no end in sight.

I love the idea of .NET Core and to have one framework to unify them all, so that you can write one code and target them all. Great.

BUT what UI framework to pick is getting more and more out of hand.

- WinForms
- WPF
- UWP
- WinUI
- Blazor
- MAUI

Maybe even VB 6.0 sooner or later when you listen to the commuinity and open source it aswell.

All those UI frameworks are mainted and kept alive, and not only kept alive with possible hot fixes/security issues but with actual features added.

So even 18 years later WinForms is mainted and some could create new applications with it.

Why does it seem impossible to let UI frameworks die? 
18 years ago .Net Framework 1.0 was born and no one (as far as I know) would even think about writing an application with this.

Don't get me wrong, this isn't me hating any of those mentioned frameworks instead I would like to understand why we can't move on and use all our resources in one UI framework like .net core is doing it.

Even the most newly added frameworks seems to overlap and there isn't the "one to rule them all" feeling like with .net core.

We are so off the path to have a unified UI framework that there is even a "hybrid" between two UI frameworks called "Blazor Mobil Bindings".

If I would image "Blazor Mobil Bindings" as a code alternative it seems like someone mixing C# and Python in one file/method, to make both C# and Python developers happy and tell them we got you covered you can use your abilities and work together.

**Comment by Happypig375** [comment]
>18 years ago .Net Framework 1.0 was born and no one (as far as I know) would even think about writing an application with this.

.NET Framework 1.0 may be dead, but .NET Framework 1.1 is still alive. You can still target `net11` in a .NET SDK-style project.

**Comment by andrewleader** [comment]
Dupe of #43, but tons of ppl agree with you (myself included). I'd voice your opinion in that thread, since there's TONS of comments and is most certainly the primary issue thread on this topic.

**Comment by nerocui** [comment]
Only WinUI here in the list is a UI framework. It can be used with both UWP and win32 APP MODELS. It is the Windows 10 implementation of Fluent Design Language. 

UWP is an app model, as oppose to win32. It's not bound to a single app framework (like WPF), so UWP can be coded in JavaScript using React Native or just plain C++. 

WPF and Win Form are app frameworks. They use the win32 app model, and they don't use composition layer for rendering. Which is why they need XAML island to render WinUI. The latest WinUI 3 desktop framework is neither WPF nor WinForm, it's simply .NET 5 Desktop with WinUI 3.

Blazor is a programming model (as oppose to MVVM, MVU) that can use any renderer(web, mobile) for UI, although the officially supported version is web.

Maui is like WPF, it's an app framework, so the UI layer is pretty flexible. Xamarin Forms now has visual for implementing different UI controls, and there's a material design visual that implements material design system in Xamarin Forms.

Picking a technology is pretty simple I think. If what you are building needs to target web, then Blazor is the only choice, you can then use whatever UI library (Material UI, Fluent UI, Bootstrap, etc) on top of that. If you need to target mobile, Maui is the only choice. Desktop going forward will all be WinUI and Project Reunion, so no confusion there either.

**Comment by mrwcjoughin** [comment]
But that’s against the whole point of Xamarin Forms - to have one
application that works on all platforms

That’s what MAUI is building apon and it needs to do better.

In the real world people don’t have massive budgets to rewrite apps in
multiple technologies

On Sat, 30 May 2020 at 03:34, nerocui <notifications@github.com> wrote:

> Only WinUI here in the list is a UI framework. It can be used with both
> UWP and win32 APP MODELS. It is the Windows 10 implementation of Fluent
> Design Language.
>
> UWP is an app model, as oppose to win32. It's not bound to a single app
> framework (like WPF), so UWP can be coded in JavaScript using React Native
> or just plain C++.
>
> WPF and Win Form are app frameworks. They use the win32 app model, and
> they don't use composition layer for rendering. Which is why they need XAML
> island to render WinUI. The latest WinUI 3 desktop framework is neither WPF
> nor WinForm, it's simply .NET 5 Desktop with WinUI 3.
>
> Blazor is a programming model (as oppose to MVVM, MVU) that can use any
> renderer(web, mobile) for UI, although the officially supported version is
> web.
>
> Maui is like WPF, it's an app framework, so the UI layer is pretty
> flexible. Xamarin Forms now has visual for implementing different UI
> controls, and there's a material design visual that implements material
> design system in Xamarin Forms.
>
> Picking a technology is pretty simple I think. If what you are building
> needs to target web, then Blazor is the only choice, you can then use
> whatever UI library (Material UI, Fluent UI, Bootstrap, etc) on top of
> that. If you need to target mobile, Maui is the only choice. Desktop going
> forward will all be WinUI and Project Reunion, so no confusion there either.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/maui/issues/122#issuecomment-636255498>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGERWXOBE7WCMPLQCEE5TRUBPCJANCNFSM4NNAXISA>
> .
>
-- 
Thanks,

Matthew Joughin
Founder / CTO

+27 81 529 7169
http://www.adaptableapps.net

**Comment by nerocui** [comment]
> But that’s against the whole point of Xamarin Forms - to have one application that works on all platforms That’s what MAUI is building apon and it needs to do better. In the real world people don’t have massive budgets to rewrite apps in multiple technologies

True.
In ideal world, there is only .Net App Framework and apps built from it runs perfectly on each and every platform. But we are not there and probably never will. Different technology exist so people can actually get work done rather than waiting for a ideal solution.

React Native exist because the sudden rise of js created a lot of js developer who can also transfer their skill to develop native apps.

Xamarin/MAUI exist because enterprise developers can apply their .NET skill to something more relevant to the mobile world.

MVVM is great and all, but people who invested in learning react or flutter may want something similar. It's not about what this thing used to be or the old vision. It's about giving devs what they want at where they want.

**Comment by Rand-Random** [comment]
Blazor Desktop enters the ring.
https://devblogs.microsoft.com/dotnet/announcing-net-6-preview-1/
https://github.com/dotnet/aspnetcore/issues/27217
